logshifter/input.go

```diff
@@ -64,24 +67,39 @@ func (input *Input) read() {
 
 		copy(cp, line)
 
-		if input.statsChannel != nil {
-			input.statsChannel <- Stat{name: "input.read", value: 1.0}
-		}
+		input.record("input.read", 1.0)
 
 		select {
 		case input.queue <- cp:
 			// queued
+			input.record("input.queue", 1.0)
 		default:
-			// evict the oldest entry to make room
-			<-input.queue
-			if input.statsChannel != nil {
-				input.statsChannel <- Stat{name: "input.drop", value: 1.0}
+			// try to evict the oldest entry to make room
+			select {
+			case <-input.queue:
+				input.record("input.evict", 1.0)
+				// try again to queue
+				select {
+				case input.queue <- cp:
+					// queued
+					input.record("input.queue", 1.0)
+				default:
```

Comment:
Is there a way this could theoretically happen? AFAICT, `(input *Input) read()` is the only code that writes to the queue, it is running in at most one thread, the queue is not closed (in `(shift *Shifter) Start()`) until the thread says it is done (in `(input *Input) Read()`), the queue size is fixed, the number of items in queue was equal to at most the size of the queue, and we just dequeued an item, so it should not be possible that writing to the queue will block a second time.

Not that I'm opposed to guarding against theoretically impossible contingencies; I'm mostly curious.